Plugin messaging Docs (#196)
PaperMC/docs: docs/paper/dev/api/plugin-messaging.md

---
slug: /dev/plugin-messaging
---

# Plugin Messaging

First introduced in [2012](https://web.archive.org/web/20220711204310/https://dinnerbone.com/blog/2012/01/13/minecraft-plugin-channels-messaging/),
Plugin messaging is a way for Paper plugins to communicate with clients. When your servers are behind a proxy, 
it will allow your Paper plugins to communicate with the proxy server.

## Sending Plugin Messages

> First, we're actually going take a look at your Paper server. Your Paper server plugin will need to register that it 
> will be sending on any given plugin channel. You'll should to do this alongside your other event listener registrations.

Reviewer: ```suggestion
First, we're going to take a look at your Paper server. Your Paper server plugin will need to register that it 
will be sending on any given plugin channel. You should to do this alongside your other event listener registrations.
```

```java
public final class PluginMessagingSample extends JavaPlugin {

    @Override
    public void onEnable() {
        getServer().getMessenger().registerOutgoingPluginChannel(this, "bungeecord:main");
        // Blah blah blah...
    }

}
```

Now that we're registered, we can send messages on the `bungeecord:main` channel.

> :::tip[The "bungeecord" and "bungeecord:main" channel]

Author: Yikes, that seems like a bit of a bug on docosaurus’ part tbh. I’d imagine it’s cause it’s looking for the colons to end the admonition, but thanks!

    @Override
    public void onEnable() {
        getServer().getPluginManager().registerEvents(this, this);
        getServer().getMessenger().registerOutgoingPluginChannel(this, "bungeecord:main");
    }

    @EventHandler
    public void onPlayerJump(PlayerJumpEvent event) {
        Player player = event.getPlayer();

        ByteArrayDataOutput out = ByteStreams.newDataOutput();
        out.writeUTF("Connect");
        out.writeUTF("hub2");
        player.sendPluginMessage(this, "bungeecord:main", out.toByteArray());
    }
  
}
```

:::tip

These channels rely on the Minecraft protocol, and are sent as a special type of packet called a 
[Plugin Message](https://wiki.vg/Plugin_channels#Plugin_Messages). They piggyback on player connections, so if a player
is not connected to the server, the server will not be able to send or receive plugin messages.

:::

### What did we just do?

We sent a plugin message on the `bungeecord:main` channel! The message we sent was a byte array that contained two strings converted to bytes: `Connect` and `hub2`.

Our proxy server received the message through the player who triggered the `PlayerJumpEvent` on our Java server. 

> Then, it recognized the channel as it's own, and in alignment with BungeeCord's format sent our player to the `hub2` server.

Reviewer: ```suggestion
Then, it recognized the channel as its own and in alignment with BungeeCord's format sent our player to the `hub2` server.
```

For BungeeCord, we can think of this message as a case-sensitive command with arguments. 
Here, our command is `Connect` and our only argument is `hub2` but some "commands" may have multiple arguments. 
For other channels introduced by client side mods, refer to their documentation to best understand how to format your messages.

### BungeeCord Plugin Message Types

> Although we sent a `Connect` message to the proxy there are a few other cases that proxies will act on. 

Reviewer: ```suggestion
Although we sent a `Connect` message to the proxy, there are a few other cases that proxies will act on. 
```

##### Example: `PlayerCount`

```java
public class MyPlugin extends JavaPlugin implements PluginMessageListener {
    
    @Override
    public void onEnable() {
        this.getServer().getMessenger().registerOutgoingPluginChannel(this, "BungeeCord");
        this.getServer().getMessenger().registerIncomingPluginChannel(this, "BungeeCord", this);
        
        Player player = ...;
        ByteArrayDataOutput out = ByteStreams.newDataOutput();
        out.writeUTF("PlayerCount");
        out.writeUTF("lobby");
        player.sendPluginMessage(this, "BungeeCord", out.toByteArray());
        // The response will be handled in onPluginMessageReceived
    }

    @Override
    public void onPluginMessageReceived(String channel, Player player, byte[] message) {
        if (!channel.equals("BungeeCord")) {
            return;
        }
        ByteArrayDataInput in = ByteStreams.newDataInput(message);
        String subchannel = in.readUTF();
        if (subchannel.equals("PlayerCount")) {
            // This is our response to the PlayerCount request
            String server = in.readUTF();
            int playerCount = in.readInt();
        }
    }
}
```

> ##### Example: `Forward`

Reviewer 2: Would it be useful to at least throw some bare bone "what exactly are you trying to do" text here?
It obviously can be understood by *reading* the code below, however these docs are meant for developers learning this. It hence might be useful to give them a heads up as to what these examples try to accomplish.

        this.getServer().getMessenger().registerIncomingPluginChannel(this, "BungeeCord", this);
        
        Player player = ...;
        ByteArrayDataOutput out = ByteStreams.newDataOutput();
        out.writeUTF("PlayerCount");
        out.writeUTF("lobby");
        player.sendPluginMessage(this, "BungeeCord", out.toByteArray());
        // The response will be handled in onPluginMessageReceived
    }

    @Override
    public void onPluginMessageReceived(String channel, Player player, byte[] message) {
        if (!channel.equals("BungeeCord")) {
            return;
        }
        ByteArrayDataInput in = ByteStreams.newDataInput(message);
        String subchannel = in.readUTF();
        if (subchannel.equals("PlayerCount")) {
            // This is our response to the PlayerCount request
            String server = in.readUTF();
            int playerCount = in.readInt();
        }
    }
}
```

##### Example: `Forward`

Forwards a plugin message to another server. This is useful for server-to-server communication within a proxy network.
For example, if a certain player is banned on one server, you can forward a message to all other servers to ban them there too.

> :::caution[Example of banning a player on all servers]

Reviewer 2: I feel like, we'd wanna move this like, behind the following example.
Reading this had me think that the following example was for banning players.

### BungeeCord Plugin Message Types

Although we sent a `Connect` message to the proxy, there are a few other cases that proxies will act on. 
These are the following:

| Message Type      | Description                                            | Arguments                                                        | Response                                          |
|:------------------|:-------------------------------------------------------|:-----------------------------------------------------------------|:--------------------------------------------------|
| `Connect`         | Connects the player to the specified server.           | `server name`                                                    | N/A                                               |
| `ConnectOther`    | Connects another player to the specified server.       | `player name`, `server name`                                     | N/A                                               |
| `IP`              | Returns the IP of the specified player.                | `player name`                                                    | `IP`, `port`                                      |

> | `PlayerCount`     | Returns the number of players on the specified server. | `server name`                                                    | `server name`, player count`                      |

Reviewer 3: The player count in the last column doesn't seems well formatted
